crates/subspace-farmer/src/piece_cache.rs

```diff
@@ -0,0 +1,666 @@
+                    let Some(offset) = cache.free_offsets.pop() else {
+                        // Not this disk farm
+                        continue;
+                    };
+
+                    if let Err(error) = cache.backend.write_piece(offset, piece_index, &piece) {
+                        error!(
+                            %error,
+                            %disk_farm_index,
+                            %piece_index,
+                            %offset,
+                            "Failed to write piece into cache"
+                        );
+                    } else {
+                        trace!(
+                            %disk_farm_index,
+                            %piece_index,
+                            %offset,
+                            "Successfully stored piece in cache"
+                        );
+                        cache.stored_pieces.insert(record_key, offset);
+                    }
+                    return;
+                }
+
+                warn!(
+                    %piece_index,
+                    "Should have inserted piece into cache, but it didn't happen, this is an \
+                    implementation bug"
+                );
+            }
+        };
+    }
+}
+
+/// Piece cache that aggregates caches of multiple disks
+#[derive(Debug, Clone)]
+pub struct PieceCache {
+    /// Individual disk caches where pieces are stored
+    caches: Arc<RwLock<Vec<DiskPieceCacheState>>>,
+    // We do not want to increase capacity unnecessarily on clone
+    worker_sender: mpsc::Sender<WorkerCommand>,
+}
+
+impl PieceCache {
+    /// Create new piece cache instance and corresponding worker.
+    ///
+    /// NOTE: Returned future is async, but does blocking operations and should be running in
+    /// dedicated thread.
+    pub fn new<NC>(node_client: NC, peer_id: PeerId) -> (Self, CacheWorker<NC>)
+    where
+        NC: NodeClient,
+    {
+        let caches = Arc::default();
+        let (worker_sender, worker_receiver) = mpsc::channel(WORKER_CHANNEL_CAPACITY);
+
+        let instance = Self {
+            caches: Arc::clone(&caches),
+            worker_sender,
+        };
+        let worker = CacheWorker {
+            peer_id,
+            node_client,
+            caches,
+            worker_receiver: Some(worker_receiver),
+        };
+
+        (instance, worker)
+    }
+
+    /// Check whether piece exists in cache
+    pub fn contains_piece(&self, key: RecordKey) -> bool {
+        // It is okay to take read lock here, writes locks are very infrequent and very short
+        for cache in self.caches.read().iter() {
+            if cache.stored_pieces.contains_key(&key) {
+                return true;
+            };
+        }
+
+        false
+    }
+
+    /// Get piece from cache
+    pub async fn get_piece(&self, key: RecordKey) -> Option<Piece> {
+        let caches = Arc::clone(&self.caches);
+
+        // TODO: Join on drop for this async task
```

Comment:
Didn't we join on drop below?

Reply:
Yes, fixed TODO before submitting PR, but forgot to remove